core/drivers/aplic_direct.c

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025 Beijing Institute of Open Source Chip (BOSC)
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <drivers/aplic.h>
+#include <drivers/aplic_priv.h>
+#include <io.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <util.h>
+
+#define APLIC_MAX_IDC            BIT(14)    /* 16384 */
+#define APLIC_IDC_BASE           0x4000
+#define APLIC_IDC_SIZE           32
+
+/* Interrupt Delivery Control (IDC) structure */
+#define APLIC_IDC_IDELIVERY      0x00
+#define APLIC_IDC_IFORCE         0x04
+#define APLIC_IDC_ITHRESHOLD     0x08
+
+#define APLIC_IDC_TOPI           0x18
+#define APLIC_IDC_TOPI_ID_SHIFT  16
+#define APLIC_IDC_TOPI_ID_MASK   GENMASK_32(25, 16)
+#define APLIC_IDC_TOPI_PRIO_MASK GENMASK_32(7, 0)
+
+#define APLIC_IDC_CLAIMI         0x1C
+
+#define APLIC_DISABLE_IDELIVERY  0
+#define APLIC_ENABLE_IDELIVERY   1
+
+#define APLIC_DISABLE_ITHRESHOLD 1
+#define APLIC_ENABLE_ITHRESHOLD  0
+
+static struct aplic_data aplic_data __nex_bss;
+
+static vaddr_t aplic_get_idc_base(void)
+{
+	struct aplic_data *aplic = &aplic_data;
+	size_t hartid = get_core_pos();
+
+	return aplic->aplic_base + APLIC_IDC_BASE + hartid * APLIC_IDC_SIZE;
+}
+
+static void aplic_set_target(struct aplic_data *aplic, uint32_t source,
+			     uint32_t hart_idx, uint32_t iprio)
+{
+	vaddr_t target = 0;
+	uint32_t val = 0;
+
+	val = SHIFT_U32(hart_idx & APLIC_TARGET_HART_IDX_MASK,
+			APLIC_TARGET_HART_IDX_SHIFT);
```

Comment:
Based on `APLIC_TARGET_HART_IDX_MASK`, the mask operation should be done after shift.
```c
	val = SHIFT_U32(hart_idx, APLIC_TARGET_HART_IDX_SHIFT) & APLIC_TARGET_HART_IDX_MASK;
```